File: 백준/Silver/1926. 그림/그림.py
from collections import deque

di = [-1, 1, 0, 0]
dj = [0, 0, -1, 1]
N ,M = map(int,input().split())
picture = []
for _ in range(N):
    picture.append(list(map(int,input().split())))
visited = [[False for _ in range(M)] for _ in range(N)]
result = []
for i in range(N):
    for j in range(M):
        if picture[i][j] == 1:
            queue = deque([[i, j]])
            cnt = 0
            while queue:
                y, x = queue.popleft()
                if visited[y][x]:
                    continue
                visited[y][x] = True
                cnt += 1
                for k in range(4):
                    new_y = y + di[k]
                    new_x = x + dj[k]
                    if 0 <= new_x < M and 0 <= new_y < N and not visited[new_y][new_x] and picture[new_y][new_x] == 1:
                        queue.append([new_y,new_x])
            if cnt > 0:
                result.append(cnt)
print(len(result))
if len(result) == 0:
    print(0)
else:
    print(max(result))

File: 백준/Silver/1926. 그림/그림.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int N, M, answer;
    static int[][] picture;
    static int[] dx = {-1,1,0,0};
    static int[] dy = {0,0,-1,1};
    static Queue<Node> q = new LinkedList<>();
    static List<Integer> result = new ArrayList<>();
    static class Node {
        int y;
        int x;
        Node(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }

    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        picture = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                picture[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                bfs(i,j);
            }
        }
        Collections.sort(result, Collections.reverseOrder());
        System.out.println(result.size());
        System.out.println(result.size() == 0 ? 0 : result.get(0));
    }

    private static void bfs(int y, int x) {
        if (picture[y][x] == 0) return;
        picture[y][x] = 0;
        q.offer(new Node(y,x));
        answer = 1;
        while (!q.isEmpty()) {
            Node tmp = q.poll();
            for (int i = 0; i < 4; i++) {
                int new_y = tmp.y + dy[i];
                int new_x = tmp.x + dx[i];
                if (0 > new_y || N <= new_y || 0 > new_x || M <= new_x
                        || picture[new_y][new_x] == 0) continue;
                picture[new_y][new_x] = 0;
                answer++;
                q.offer(new Node(new_y, new_x));
            }
        }
        result.add(answer);
    }
}

